### PR TITLE
#307 CreateView implementation

### PR DESCRIPTION
**Pull request by vryazanov**
#307 
* CreateView implemented
* tests fixed

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#369](https://github.com/timelesslounge/timelessis/issues/369) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#369](https://github.com/timelesslounge/timelessis/issues/369) is assigned to @vladarefiev[/z](https://www.0crat.com/u/vladarefiev), here is [why](https://www.0crat.com/footprint/CF9LG1YCA/b13766bb-b4a9-4de6-8a65-b843e150d335); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: timeless/restaurants/models.py [hunk 1]
```diff
@@ -113,19 +113,11 @@ class Table(PosterSyncMixin, DB.Model):
...
     min_capacity = DB.Column(DB.Integer, DB.ForeignKey("scheme_types.id"))
     deposit_hour = DB.Column(DB.Integer, DB.ForeignKey("scheme_types.id"))
-    created = DB.Column(DB.DateTime, nullable=False)
-    updated = DB.Column(DB.DateTime, nullable=False)
 
     reservations = DB.relationship("TableReservation", back_populates="table")
 
     DB.UniqueConstraint(u"name", u"floor_id")
 
-    @validate_required("name", "x", "y", "width", "height", "status",
```
Comment (vladarefiev) [inline timeless/restaurants/models.py]: @vryazanov We do not need this anymore ?

File: timeless/restaurants/tables/forms.py [hunk 2]
```diff
@@ -2,7 +2,7 @@
 from timeless.restaurants import models
 
 
-class TableForm(forms.BaseModelForm):
+class TableForm(forms.ModelForm):
```
Comment (vladarefiev) [inline timeless/restaurants/tables/forms.py]: @vryazanov Tell me pls what is difference?

File: timeless/views.py [hunk 3]
```diff
@@ -224,6 +225,51 @@ def get_default_context(self):
         return context
 
 
+class CreateView(GenericView):
+    """ Class which creates objects based on received POST data and provided
+    form class """
+    form_class = None
+    success_view_name = None
```
Comment (vladarefiev) [inline timeless/views.py]: @vryazanov Hmm, I think it's actually not view name it's url name. wdyt?

File: timeless/views.py [hunk 4]
```diff
@@ -224,6 +225,51 @@ def get_default_context(self):
...
+
+    def get_form(self, *args, **kwargs):
+        """ Create form instance """
+        if self.form_class is None:
```
Comment (vladarefiev) [inline timeless/views.py]: @vryazanov why not just `if not self.form_class` ?

File: timeless/views.py [hunk 5]
```diff
@@ -224,6 +225,51 @@ def get_default_context(self):
...
+
+        if not form.validate():
+            return self.render_to_response(self.get_context(form=form))
+
+        form.save()
+        return redirect(self.get_success_url_redirect())
+
+
+class UpdateView(GenericView):
+    """ Base view for updating objects"""
```
Comment (vladarefiev) [inline timeless/views.py]: @vryazanov leave puzzles for continue pls

File: timeless/views.py [hunk 6]
```diff
@@ -224,6 +225,51 @@ def get_default_context(self):
...
+    def get_form(self, *args, **kwargs):
+        """ Create form instance """
+        if self.form_class is None:
+            raise NotImplementedError(f"{self.__class__.__name__} must define "
+                                      f"'form_class' attribute")
+        return self.form_class(*args, **kwargs)
+
+    def get_success_url_redirect(self):
+        """ Reverse URL based on view name """
+        if self.success_view_name is None:
```
Comment (vladarefiev) [inline timeless/views.py]: @vryazanov the same

**Comment by vladarefiev** [comment]
@vryazanov looks greate, thanks. just a couple of comments

File: timeless/restaurants/tables/forms.py (hunk 2 above)
Comment (vryazanov) [reply timeless/restaurants/tables/forms.py]: @vladarefiev `ModelForm` can `create` / `update` instances

File: timeless/restaurants/models.py (hunk 1 above)
Comment (vryazanov) [reply timeless/restaurants/models.py]: @vladarefiev No, we don't. It breaks logic of forms. Form at first creates an instance without any args in constructor (and on this step validator raises an exception) and only then sets attributes.

File: timeless/views.py (hunk 3 above)
Comment (vryazanov) [reply timeless/views.py]: @vladarefiev I thought about it, but then decided that view name is better then hardcode url

File: timeless/views.py (hunk 5 above)
Comment (vryazanov) [reply timeless/views.py]: @vladarefiev I aleady have jobs in my agenda related to `DeleteView` and `UpdateView`, so I will implement them

**Comment by vryazanov** [comment]
@vladarefiev I corrected the code and answered to your questions. Please have a look

**Comment by vladarefiev** [comment]
@rultor merge

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by 0crat** [comment]
Job was finished in 13 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @vladarefiev[/z](https://www.0crat.com/u/vladarefiev), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#369](https://github.com/timelesslounge/timelessis/issues/369) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa)

**Comment by ypshenychka** [comment]
@vladarefiev According to our [QA Rules](http://www.zerocracy.com/policy.html#42):
>The code reviewer found at least three problems in the code.
Comments were mostly about design problems, not cosmetic issues.

Only two issues were found during code review. 
Please confirm that you'll try to find at least three major problems while future reviews.

**Comment by vladarefiev** [comment]
@ypshenychka confirm

**Comment by ypshenychka** [comment]
@vladarefiev thanks

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +20 point(s) just awarded to @vladarefiev[/z](https://www.0crat.com/u/vladarefiev)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Pull request merged**